### PR TITLE
CLOUDSTACK-9348: NioConnection improvements

### PR DESCRIPTION
**Pull request by rhtyd**
Reopened PR with squashed changes for a re-review and testing after https://github.com/apache/cloudstack/pull/1493 and sub-sequent PRs got reverted

### CODE REVIEW

**Comment by rhtyd** [comment]
@swill I was able to reproduce the addHost issue and fixed it. I was able to deploy against a KVM env with two clusters with Ubuntu and CentOS based hosts in each: (addHost can fail silently if we're trying to mix CentOS and Ubuntu KVM hosts in the same cluster)

screenshot from 2016-05-19 16-01-48

If possible, I would request for this bugfix to be added in 4.9.0 release as it fixes:

- critical nio reconnection, improves SSL handling by making it non-blocking
- without this fix, agents (kvm, systemvm etc) can take some time (minutes to hours) to reconnect
- the management server's agent handler on port 8250 can be pwned by a single telnet (or similar malicious) client. To test, on existing deployments without this patch just connect a telnet agent on port 8250 of the mgmt server, now this mgmt server can no longer immediately handle any agent connection (try restarting kvm agent for example).

**Comment by swill** [comment]
I will get this run through CI.  Just to be clear, this PR includes all of the changes that were included in 1543, 1538, 1534, 1493 as well as a fix for the `addHost` issue.  Is that correct?  Thx...

**Comment by rhtyd** [comment]
@swill yes, this PR includes all the reverted changes and the addHost fix.

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 0
   Errors: 0
 Duration: 8h 51m 55s
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_19_2016_19_29_38_NAMX3Q:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_19_2016_19_29_38_NAMX3Q/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_19_2016_19_29_38_NAMX3Q/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_19_2016_19_29_38_NAMX3Q/runinfo.txt)

**`/tmp/MarvinLogs/test_network_V1CHD4:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_V1CHD4/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_V1CHD4/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_V1CHD4/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_L9ZGGS:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_L9ZGGS/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_L9ZGGS/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_L9ZGGS/runinfo.txt)


Uploads will be available until `2016-07-20 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
I am going to run the test again because it was an intermittent issue, so a couple tests is not a bad idea.

**Comment by rhtyd** [comment]
Thanks @swill

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 83
  Skipped: 0
   Failed: 0
   Errors: 2
 Duration: 8h 27m 27s
```

**Summary of the problem(s):**
```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcRemoteAccessVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 293, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_AVQ8UF/results.txt
```

```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcSite2SiteVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 472, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_AVQ8UF/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_20_2016_06_41_21_MDBMOC:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_20_2016_06_41_21_MDBMOC/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_20_2016_06_41_21_MDBMOC/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/DeployDataCenter__May_20_2016_06_41_21_MDBMOC/runinfo.txt)

**`/tmp/MarvinLogs/test_network_AVQ8UF:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_AVQ8UF/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_AVQ8UF/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_network_AVQ8UF/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_JFF9WZ:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_JFF9WZ/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_JFF9WZ/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1549/tmp/MarvinLogs/test_vpc_routers_JFF9WZ/runinfo.txt)


Uploads will be available until `2016-07-20 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
Two clean runs in a row.  I think this one is ready to put back in now...

**Comment by swill** [comment]
BTW, thank you for your quick response time to work through issues on your PRs and get things stabilized.  I really do appreciate it.

**Comment by rhtyd** [comment]
@swill thank you for your time in testing this

**Comment by swill** [comment]
And of course the first build I do after I merge it into master I get the `addHost` error and I did not get it once after I pulled it out.  Haha...  I will build again and see...

**Comment by swill** [comment]
@rhtyd can you specify the logs you need if I run into this issue so you can track down why it is happening?  I will leave it in for now but I would like to send you the logs every time there is an issue so you can understand what is going on so we can try to fix it (once and for all).  Thanks...

**Comment by swill** [comment]
@rhtyd uggg...  The first two builds on master after merging this back in have both resulted in the `addHost` error.  :(  Let me know what you need to troubleshoot this.  I have a couple PRs I want to do a final round of testing on, so I will try to get them tested and we will see if they also run into this issue when I try to test them.  If they do, I may have to revert again.  I am trying my best to get a bunch of the PRs that are close in, so I am pushing the freeze date a bit (obviously).  I will probably freeze over the weekend or early next week.  I would like to make sure we have this resolved though.  I don't want this to derail my final testing.  I will launch a bunch more CI right now and we will see how things go...

**Comment by rhtyd** [comment]
@swill if it's failing for addHost, can you send the mgmt server log, marvin log and the agent log, and details of the environment (as much as possible). I can have a look at it tonight or tomorrow morning. If it helps we can pair-debug the issue over GTM in your CI environment?

It would also help if anyone else is able to reproduce the error on current master as well, pinging @DaanHoogland @wido @nlivens @nvazquez @koushik-das @kiwiflyer and others.

@swill if we're unable to fix the issue then sure we can revert it.

**Comment by rhtyd** [comment]
@swill I ran tests on latest master using packages: http://packages.shapeblue.com/cloudstack/custom/master-latest/

I was able to deploy a zone with two KVM hosts and a mgmt server (on centos6). I was able to use a cloudmonkey based script and it was able to add multiple hosts without any error; here is the log and deploy script:

[cloudmonkey.log.txt](https://github.com/apache/cloudstack/files/276870/cloudmonkey.log.txt)
[deploy-kvm-adv.sh.txt](https://github.com/apache/cloudstack/files/276871/deploy-kvm-adv.sh.txt)

@swill I was then able to launch VMs, see systemvm connect and perform regression tests. I'm unable to reproduce the issue now, I think master is stable; in case you hit the `addHost` issue please send me the mgmgt server and agent logs. Thanks.

**Comment by swill** [comment]
@rhtyd fingers crossed, things seem to be in pretty good shape.  The first two runs after the merge had `addHost` errors, but I have done about 6-7 CI runs since and have not gotten the issue again, so things seem to be pretty good.  Thanks for working on this...

**Comment by swill** [comment]
I think I am going to revert this PR.  We are having intermittent issues with it still and I am not confident running a production environment with this in place at this time, so I don't think I can justify leaving it in without us doing some more testing to figure out what is going on.

I have had a few reports like this as people test 4.9, which are concerning:
> NIO SSL agent not connecting. when I telnet to 8250, the agent immediately came up without me having to restart it.

I am also still periodically getting the `addHost` issue we thought we had resolved previously.

After I revert his, can you create a new PR with this same code so we can start getting more concrete testing on it and we can start consolidating some logs when it misbehaves?

**Comment by rhtyd** [comment]
@swill may I have the mgmt server and agent logs when the failures were intercepted. This is to make sure it's not your environment specific issue. I'll also need the JRE version in use (openjdk, or oraclejdk, which versions specifically?). If possible can you also take a heapdump and share that with me (run jmap -dump:file=heap.bin <pid of the java process>, gzip and scp this bin file and please share this somewhere for both mgmt server and agent).

"NIO SSL agent not connecting. when I telnet to 8250, the agent immediately came up without me having to restart it." -- this is something which I've fixed in latest master (using timeout on selectors), can you ask them if they are using latest master?

We've seen this fix deployed in a very large environment with 1000s of hosts and I've not heard anything from them. We've not gotten any reported on MLs so far, I would appreciate if those people who are experiencing issues can share it on public channels. Thanks.

File: utils/src/main/java/com/cloud/utils/nio/NioConnection.java [hunk 1]
```diff
@@ -125,7 +125,7 @@ public boolean isStartup() {
     public Boolean call() throws NioConnectionException {
         while (_isRunning) {
             try {
-                _selector.select();
+                _selector.select(1000);
```
Comment (rhtyd) [inline utils/src/main/java/com/cloud/utils/nio/NioConnection.java]: @swill this change ^^ causes the selector loop to never block indefinitely but at most 1 second. This ensures that reconnections are fast. If anyone experiences the behaviour that tel-netting to port 8250 causes agent to reconnect then, either (1) they are not using latest master or (2) they did not wait for up to one second (i.e. hit an edge case), in which case we can lower this number to 100-500 millisecond at the expense of increasing cpu usage.

**Comment by swill** [comment]
This was shared in the "4.9/master Testing Coordination" thread.  Simon Weller and his team at ENA have run into this on the latest master in both of their hardware labs while testing master for me in preparation for the 4.9 RC.

I am still getting the `addHost` issue periodically which showed up a lot when this had bigger issues.

I am concerned about this in production to be honest.  How long has this been running in production at your client?

**Comment by rhtyd** [comment]
@swill close to two months now. Can you add the mgmt server logs, heap dump and any other dumps which can help me fix the issue.
@swill @kiwiflyer Can you please open a JIRA issue where you can put these details, without necessary information we cannot find and fix the issue or conclude that the issue was caused by something else. I've built latest master repository here: http://packages.shapeblue.com/cloudstack/custom/testing
I'll continue this discussion on the ML thread.

**Pull request merged**